src/Symfony/Component/DependencyInjection/Compiler/CheckTypeHintsPass.php

```diff
@@ -0,0 +1,165 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Exception\InvalidParameterTypeHintException;
+
+/**
+ * Checks whether injected parameters types are compatible with type hints.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Julien Maulny <jmaulny@darkmira.fr>
+ */
+class CheckTypeHintsPass extends AbstractRecursivePass
+{
+    private $autoload;
+
+    public function __construct(bool $autoload = false)
+    {
+        $this->autoload = $autoload;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        // When service is declared with $container->register(Foo::class); and not $container->register(Foo::class, Foo::class);
+        if (null === $value->getClass()) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (!$this->autoload && !\class_exists($this->getClassName($value), false)) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (null !== $constructor = $this->getConstructor($value, false)) {
+            $this->checkArgumentsTypeHints($constructor, $value->getArguments());
+        }
+
+        foreach ($value->getMethodCalls() as $methodCall) {
+            $reflectionMethod = $this->getReflectionMethod($value, $methodCall[0]);
+
+            $this->checkArgumentsTypeHints($reflectionMethod /* warn */, $methodCall[1]);
+        }
+
+        return parent::processValue($value, $isRoot);
+    }
+
+    /**
+     * Check type hints for every parameter of a method/constructor.
+     *
+     * @param \ReflectionMethod $reflectionMethod
```

Comment:
should be removed: we don't add annotations when they just duplicate the signature (same everywhere in the PR)